### PR TITLE
[Core] Adapted BuildLHS

### LINKED ISSUE #7022: BlockBuilder builds RHS in BuildLHS
https://github.com/KratosMultiphysics/Kratos/blob/0f7b1083807041af6f8c927cd518e5be9dad9987/kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h#L301-L312

to be refactored

the TrilinosBlockBuilder does this correct

FYI @emiroglu

### PR DESCRIPTION
**Pull request by loumalouomega**
Fixes #7022

### CODE REVIEW

**Comment by philbucher** [comment]
Thx, goes into the right direction 

I would say that Build should internally call BuildLHS and BuildRHS to avoid code duplication or am I missing sth?

**Comment by loumalouomega** [comment]
> 
> 
> Thx, goes into the right direction
> 
> I would say that Build should internally call BuildLHS and BuildRHS to avoid code duplication or am I missing sth?

It is more efficient this way, despite "duplicate code". There are different calls if you just ask the element/condition for the LHS, RHS or LHS+RHS and therefore to avoid additional computations this is, despite its duplications, the most efficient way

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 1]
```diff
@@ -301,12 +301,83 @@ class ResidualBasedBlockBuilderAndSolver
...
+                    AssembleLHS(rA, lhs_contribution, equation_id);
+
+                    // Clean local elemental memory
+                    pScheme->CleanMemory(*it_cond);
+                }
+            }
+        }
+
+        const double stop_build = OpenMPUtils::GetCurrentTime();
+        KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolver", (this->GetEchoLevel() >= 1 && rModelPart.GetCommunicator().MyPID() == 0)) << "Build time: " << stop_build - start_build << std::endl;
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]:
```suggestion
        KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolver", this->GetEchoLevel() >= 1) << "Build time LHS: " << stop_build - start_build << std::endl;

```

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 2]
```diff
@@ -301,12 +301,83 @@ class ResidualBasedBlockBuilderAndSolver
...
+
+        KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolver", (this->GetEchoLevel() > 2 && rModelPart.GetCommunicator().MyPID() == 0)) << "Finished parallel building" << std::endl;
```
Comment (philbucher) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h]:
```suggestion
        KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolver", this->GetEchoLevel() > 2) << "Finished parallel building LHS" << std::endl;

```

**Review by philbucher — COMMENTED** [review]
looks ok to me, thx

@emiroglu could you please test if this works for you?

@RiccardoRossi do you also want to take a look?

**Comment by emiroglu** [comment]
> looks ok to me, thx
> 
> @emiroglu could you please test if this works for you?
> 
> @RiccardoRossi do you also want to take a look?

Yes this works fine

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h [hunk 3]
```diff
@@ -1590,6 +1663,21 @@ class ResidualBasedBlockBuilderAndSolver
         }
     }
 
+    //**************************************************************************
```
Comment (RiccardoRossi) [inline kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:1666]: I think this function belongs to the space...

Comment (philbucher) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:1666]: There are already several custom fcts like this
If this should go to the space I would do it separately and with caution 
Whoever wrote those fcts probably had a reason not to use the already existing one in the space

Comment (loumalouomega) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:1666]: I think it is not the same in the block and the residual, in any case I think for now this should stay in here

**Review by philbucher — APPROVED** [review]
please wait until Friday in case there are more comments

**Comment by loumalouomega** [comment]
> please wait until Friday in case there are more comments

imagen

File: kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h (hunk 3 above)
Comment (RiccardoRossi) [reply kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h:1666]: ok

**Review by RiccardoRossi — COMMENTED** [review]
this looks good to me

**Comment by philbucher** [comment]
since everyone is ok, I don't see the need to wait, please go ahead :)

**Pull request merged**